glfs.c

```diff
@@ -580,6 +581,7 @@ static void glfs_async_cbk(glfs_fd_t *fd, ssize_t ret, void *data)
 			break;
 		case TCMU_GLFS_WRITE:
 		case TCMU_GLFS_FLUSH:
+		case TCMU_GLFS_DISCARD:
 			ret =  tcmu_set_sense_data(cmd->sense_buf, MEDIUM_ERROR,
 			                           ASC_WRITE_ERROR, NULL);
```

Comment:
Do you guys have any specific limits on size of alignment? If so call tcmu_set_dev_opt_unmap_gran/tcmu_set_dev_unmap_gran_align in your handler's open() call to override the defaults.

Also, do you have any special error codes that can be returned for discard like if failed due to a resize operation going on? If so we would want to handle it here by returning a different sense code.

Reply:
@mikechristie glfs_discard, eventually calls fallocate with FALLOC_FL_PUNCH_HOLE on the brick, so you can specify any valid offset and length and it will do the right thing. Just adding more info from man page, just in case you have any insight about it based on that.
Deallocating file space
       Specifying  the  FALLOC_FL_PUNCH_HOLE  flag (available since Linux 2.6.38) in mode deallocates space (i.e., creates a hole) in the byte range starting at offset and continuing for
       len bytes.  Within the specified range, partial filesystem blocks are zeroed, and whole filesystem blocks are removed from the file.  After a  successful  call,  subsequent  reads
       from this range will return zeroes.

@mikechristie I am not sure what should be the behaviour with parallel operations. Resize operation in glfs, eventually calls either a truncate/zerofill (based on sparse or not, zerofill is a form of fallocate()). Both these operations take a lock on the file to make sure one operation happens after the other. Could you help us understand the case where we may want to give an error for parallel operations?

Reply:
For shrink, the resize operation and discard be sent in parallel from runner's perspective, so the discard could by pass the offset and len checks. If glfs.c sent the discard after you did the resize, then would xfs's fallocate call send an error like EFBIG?

We are not supporting shrink in rbd.c so we never handled it in there.

For the zero'ing issue offlist, I think you will be ok.

Reply:
@mikechristie I did some tests with fallocate -p with offset larger then size of the file and it kept succeeding. Let me do more tests and confirm.

Reply:
Hey,

So was this patch good to go on your side as far as testing goes?

Reply:
@mikechristie Sorry for the delay. Was busy with some other work. Got the results now.

22:16:18 :( ⚡ ls -l /mnt/r3/
total 0
-rw-r--r--. 1 root root 0 May 17 22:15 a

Got the strace and saw that fallocate with offset, length 1024 which is greater than the size of the file (i.e. 0) is succeeding. So we are good to go here

strace.8444:fallocate(19, FALLOC_FL_KEEP_SIZE|FALLOC_FL_PUNCH_HOLE, 1024, 1024) = 0 <0.000011>